pkg/handlers/generic/lifecycle/cosi/handler.go

```diff
@@ -0,0 +1,192 @@
+		config:              cfg,
+		helmChartInfoGetter: helmChartInfoGetter,
+		variableName:        v1alpha1.ClusterConfigVariableName,
+		variablePath:        []string{"addons", v1alpha1.COSIVariableName},
+	}
+}
+
+func (n *DefaultCOSIController) Name() string {
+	return "COSIControllerHandler"
+}
+
+func (n *DefaultCOSIController) AfterControlPlaneInitialized(
+	ctx context.Context,
+	req *runtimehooksv1.AfterControlPlaneInitializedRequest,
+	resp *runtimehooksv1.AfterControlPlaneInitializedResponse,
+) {
+	commonResponse := &runtimehooksv1.CommonResponse{}
+	n.apply(ctx, &req.Cluster, commonResponse)
+	resp.Status = commonResponse.GetStatus()
+	resp.Message = commonResponse.GetMessage()
+}
+
+func (n *DefaultCOSIController) BeforeClusterUpgrade(
+	ctx context.Context,
+	req *runtimehooksv1.BeforeClusterUpgradeRequest,
+	resp *runtimehooksv1.BeforeClusterUpgradeResponse,
+) {
+	commonResponse := &runtimehooksv1.CommonResponse{}
+	n.apply(ctx, &req.Cluster, commonResponse)
+	resp.Status = commonResponse.GetStatus()
+	resp.Message = commonResponse.GetMessage()
+}
+
+func (n *DefaultCOSIController) apply(
+	ctx context.Context,
+	cluster *clusterv1.Cluster,
+	resp *runtimehooksv1.CommonResponse,
+) {
+	clusterKey := ctrlclient.ObjectKeyFromObject(cluster)
+
+	log := ctrl.LoggerFrom(ctx).WithValues(
+		"cluster",
+		clusterKey,
+	)
+
+	varMap := variables.ClusterVariablesToVariablesMap(cluster.Spec.Topology.Variables)
+
+	cosiVar, err := variables.Get[apivariables.COSI](varMap, n.variableName, n.variablePath...)
+	if err != nil {
+		if variables.IsNotFoundError(err) {
+			log.
+				Info(
+					"Skipping COSI handler, cluster does not specify request COSI Controller addon deployment",
+				)
+			return
+		}
+		log.Error(
+			err,
+			"failed to read COSI variable from cluster definition",
+		)
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage(
+			fmt.Sprintf("failed to read COSI variable from cluster definition: %v",
+				err,
+			),
+		)
+		return
+	}
+
+	var strategy addons.Applier
+	switch ptr.Deref(cosiVar.Strategy, "") {
+	case v1alpha1.AddonStrategyHelmAddon:
+		helmChart, err := n.helmChartInfoGetter.For(ctx, log, config.COSIController)
+		if err != nil {
+			log.Error(
+				err,
+				"failed to get configmap with helm settings",
+			)
+			resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+			resp.SetMessage(
+				fmt.Sprintf("failed to get configuration to create helm addon: %v",
+					err,
+				),
+			)
+			return
+		}
+		strategy = addons.NewHelmAddonApplier(
+			n.config.helmAddonConfig,
+			n.client,
+			helmChart,
+		)
+	case v1alpha1.AddonStrategyClusterResourceSet:
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage(
+			fmt.Sprintf(
+				"strategy %q not provided for COSI", v1alpha1.AddonStrategyClusterResourceSet,
+			),
+		)
+		return
+	case "":
+		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
+		resp.SetMessage("strategy not provided for COSI")
+		return
```

Comment:
why is a "empty" strategy a separate clause from default?

Reply:
This is the pattern from other handlers, mainly just the error is a little clearer to differentiate a wrong strategy vs no strategy.

In a normal flow, it should never get to either of these cases since the API enforces the supported cases.